homeassistant/components/jewish_calendar/sensor.py

```diff
@@ -104,6 +112,41 @@ def __init__(
         self.havdalah_offset = havdalah_offset
         _LOGGER.debug("Sensor %s initialized", self.type)
 
+        def recalculate(time=0):
+          _LOGGER.debug("Sunset with offset Triggred")
+          hass.loop.create_task(self.async_update())
+
+        def recalculate_and_schedule():
+          _LOGGER.debug("Sunset without offset Triggred")
+          hass.loop.create_task(self.async_update())
+          schedule_havdalah_recalculate()
+
+        def schedule_havdalah_recalculate():
+          today = dt_util.as_local(dt_util.now()).date()
+          if self.havdalah_offset == 0:
+            location = hdate.Location(latitude=self.latitude,
+                                  longitude=self.longitude,
+                                  timezone=self.timezone,
+                                  diaspora=self.diaspora)
+            times = hdate.Zmanim(
+                today, location, self.candle_lighting_offset,
+                self.havdalah_offset, self._hebrew)
+            next_schedule = dt_util.as_local(times.zmanim["three_stars"])
+          else:
+            sunset = dt_util.as_local(get_astral_event_date(
+              self.hass, SUN_EVENT_SUNSET, today))
+            next_schedule = sunset+dt.timedelta(minutes=self.havdalah_offset)
+          _LOGGER.debug("Will also recalculate at %s", str(next_schedule))
+          now = dt_util.as_local(dt_util.now())
+          recalculate_in_seconds = int((next_schedule-now).total_seconds())+1
+          async_call_later(self.hass, recalculate_in_seconds, recalculate)
+        # We always give 1 extra second
+        async_track_sunset(self.hass, recalculate, -dt.timedelta(minutes=self.candle_lighting_offset-1,seconds=59))
+        async_track_sunset(self.hass, recalculate_and_schedule, dt.timedelta(seconds=1))
+        async_track_time_change(self.hass, recalculate, hour=0, minute=0, second=1)
+        # Just in case we got started between sunset and havdalah
+        schedule_havdalah_recalculate()
+
```

Comment:
Please see the sun component where this kind of scheduling has been implemented. Simply reuse at the end of the async_update method the value for when the sensor is supposed to be updated.

Reply:
In the Sun component it's not a set time every day, and it should only be scheduled for the next event after the current event has passed, vs here we have a predefined schedule
1.  At candle lighting
2.  At Sunset
3.  At Havdalah
4.  Midnight - I'm not sure this is required at all since should get the next day from the sunset calculation.

Reply:
I understand, but I still think it would be wiser instead of adding a whole bunch of calculations to pre-build a schedule, simply to schedule the next update at the end of the current update.

Now that I better understand the code, I think your code has a major problem as the `__init__` method which will calculate the schedule will base itself on sunset, candle lighting, and Havdalah at the time that home assistant has been started. But these values are dynamic so the best way I think would be to in the update method to get all the times no matter if they are needed for setting the state. Then they can be sorted and once the next time has arrived, we can get the next time. Or something of the kind.

Reply:
@tsvi 
My goal here was to reduce the calculations to a minimum that will do the job, by delegating the scheduling to the system which already have an offset built in, I saw it as a the way to go, I can't believe that the systems sunset trigger would not account for any changes that you are concerned about.

This is valid for candle lighting and sunset, but the challenge was for havdalah if using the default which is three_stars, so since we only have to trigger that once a day and the update method is called multiple times a day, we have 2 ways of doing it either figure out within the method if at this iteration we need to schedule a havdalah calculation, or just schedule it outside that method, I elected the latter, but either way will work.

Reply:
Did you actually check how much your code saves in compute time?

The way your code is written seems to me to create a huge burden on code maintainability which I think would be worse.

Thank you for your effort and for pointing out the issue. But I'll need you to rewrite your code to be more maintainable and then I'll take another look at the PR.

Reply:
Thanks so much @tsvi, I'm no python expert, I would be very happy if you do it in any better way.

Just another note while your working on rewriting it, I would suggest, to throttle the calculation by saving an instance of zmanim in hass.data and looking up the values for each sensor rather than recalculating for each sensor.